impeller/entity/entity_pass.cc

```diff
@@ -321,13 +325,23 @@ bool EntityPass::Render(ContentContext& renderer,
                   Rect::MakeSize(root_render_target.GetRenderTargetSize()),
                   {.readonly = true});
 
-  IterateAllEntities([lazy_glyph_atlas =
-                          renderer.GetLazyGlyphAtlas()](const Entity& entity) {
-    if (const auto& contents = entity.GetContents()) {
-      contents->PopulateGlyphAtlas(lazy_glyph_atlas, entity.DeriveTextScale());
-    }
-    return true;
-  });
+  int32_t required_mip_count = 1;
+  IterateAllElements(
+      [&required_mip_count, lazy_glyph_atlas = renderer.GetLazyGlyphAtlas()](
+          const Element& element) {
+        if (auto entity = std::get_if<Entity>(&element)) {
+          if (const auto& contents = entity->GetContents()) {
+            contents->PopulateGlyphAtlas(lazy_glyph_atlas,
+                                         entity->DeriveTextScale());
+          }
+        }
+        if (auto subpass = std::get_if<std::unique_ptr<EntityPass>>(&element)) {
+          const EntityPass* entity_pass = subpass->get();
+          required_mip_count =
+              std::max(required_mip_count, entity_pass->GetRequiredMipCount());
+        }
+        return true;
+      });
 
```

Comment:
I don't like that we're adding on to this callback - this is a fairly expensive traversal that we're doing in addition to the regular traversal. Can we compute this during normal traversal?

Reply:
For example, here:

https://github.com/flutter/engine/blob/main/impeller/entity/entity_pass.cc#L531-L551

Or here:

https://github.com/flutter/engine/blob/main/impeller/entity/entity_pass.cc#L856-L874

Reply:
This lookahead traversal already exists, this just adds one extra branch for looking at mip counts.  So the addition won't be expensive.  The problem is that the resource is created before we actually render it, so there has to be some form of look ahead afaik.  I didn't see an alternative, but maybe brandon has a suggestion.

Reply:
Yes but I want to remove this lookahead, and this makes it harder :)

Reply:
I mean, you have to either look ahead, or be able to switch mid render.  Since mip map counts have to be specified at creation I think that would mean assuming mip count 1, then blitting to a mipmap texture if we ever encounter something that needs mipmaps.

The other option is to unconditionally add mipmaps which we said we don't want.

Reply:
Feel free to file a bug for this and I'll fix it later then

Reply:
Okay, let's see what @bdero says.  Maybe there is another lookahead already happening I can attach this logic to, so the one I'm adding to can get ripped out at some point still.  Otherwise I'll file an issue.

Reply:
We can work this out as the Aiks dispatch unfolds. Added some details here: https://github.com/flutter/engine/pull/49607#discussion_r1450912964

Reply:
Thanks brandon.  Storing it on the pass instead of the subpass makes sense.  I'll file an issue to follow up when this gets approval.

Reply:
Issue filed: https://github.com/flutter/flutter/issues/141483

Reply:
This implementation is wrong:
* You'd need to perform this look-ahead on each EntityPass. `Render` is only called for the root pass. (Notice that the other call to `CreateRenderTarget` always has a mip count of 1).
* `IterateAllElements` doesn't iterate through only direct children, it loops through the entire element tree recursively.

Reply:
So, if I have nested save layers the render target won't have mip maps if it can't be collapsed?  Something like this?

```c++
Canvas canvas;
canvas.Save({.blend_mode = kMultiply});
canvas.DrawCircle();
canvas.Save({}, ImageFilter::MakeBlur());
```

Reply:
Yeah. Inverting this to remove the nested `SaveLayer` would also have a similarly mip-less result:

```cpp
Canvas canvas;
canvas.SaveLayer({.image_filter = ImageFilter::MakeBlur()});
canvas.DrawCircle();
```

Reply:
The following works with the proper mipmaps, is this what you had in mind?
```
TEST_P(AiksTest, GaussianBlurFoo) {
  Canvas canvas;
  canvas.DrawCircle({100, 100}, 50, {.color = Color::CornflowerBlue()});
  canvas.SaveLayer({}, std::nullopt,
                   ImageFilter::MakeBlur(Sigma(30), Sigma(30),
                                         FilterContents::BlurStyle::kNormal,
                                         Entity::TileMode::kClamp));
  canvas.DrawCircle({200, 200}, 50, {.color = Color::Chartreuse()});

  ASSERT_TRUE(OpenPlaygroundHere(canvas.EndRecordingAsPicture()));
}
```

Reply:
This gets us in a place without mipmaps:
```c++
TEST_P(AiksTest, GaussianBlurFoo) {
  Canvas canvas;
  canvas.DrawPaint({.color = Color::Wheat()});
  canvas.SaveLayer({.blend_mode = BlendMode::kMultiply});
  canvas.DrawCircle({100, 100}, 50, {.color = Color::CornflowerBlue()});
  canvas.SaveLayer({}, std::nullopt,
                   ImageFilter::MakeBlur(Sigma(30), Sigma(30),
                                         FilterContents::BlurStyle::kNormal,
                                         Entity::TileMode::kClamp));
  canvas.DrawCircle({200, 200}, 50, {.color = Color::Chartreuse()});

  ASSERT_TRUE(OpenPlaygroundHere(canvas.EndRecordingAsPicture()));
}
```

Reply:
> is this what you had in mind?

The example I gave sets an image filter on the paint state handed to the `SaveLayer` -- no backdrop filter involved. So something like:

```cpp
TEST_P(AiksTest, GaussianBlurFoo) {
  Canvas canvas;
  canvas.SaveLayer({.image_filter = ImageFilter::MakeBlur(...)});
  canvas.DrawCircle({200, 200}, 50, {.color = Color::Chartreuse()});

  ASSERT_TRUE(OpenPlaygroundHere(canvas.EndRecordingAsPicture()));
}
```